File: src/main/java/com/github/spotbugs/snom/SpotBugsTask.java
/**
 * Copyright 2019 SpotBugs team
 *
 * <p>Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
 * except in compliance with the License. You may obtain a copy of the License at
 *
 * <p>http://www.apache.org/licenses/LICENSE-2.0
 *
 * <p>Unless required by applicable law or agreed to in writing, software distributed under the
 * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.spotbugs.snom;

import com.github.spotbugs.snom.internal.SpotBugsHtmlReport;
import com.github.spotbugs.snom.internal.SpotBugsTextReport;
import com.github.spotbugs.snom.internal.SpotBugsXmlReport;
import edu.umd.cs.findbugs.annotations.NonNull;
import edu.umd.cs.findbugs.annotations.Nullable;
import edu.umd.cs.findbugs.annotations.OverrideMustInvoke;
import groovy.lang.Closure;
import java.io.File;
import java.util.Collection;
import java.util.HashSet;
import java.util.Set;
import java.util.stream.Collectors;
import org.gradle.api.Action;
import org.gradle.api.DefaultTask;
import org.gradle.api.GradleException;
import org.gradle.api.InvalidUserDataException;
import org.gradle.api.NamedDomainObjectContainer;
import org.gradle.api.artifacts.Configuration;
import org.gradle.api.file.FileCollection;
import org.gradle.api.model.ObjectFactory;
import org.gradle.api.provider.ListProperty;
import org.gradle.api.provider.Property;
import org.gradle.api.provider.Provider;
import org.gradle.api.reporting.SingleFileReport;
import org.gradle.api.tasks.Input;
import org.gradle.api.tasks.InputFile;
import org.gradle.api.tasks.InputFiles;
import org.gradle.api.tasks.Internal;
import org.gradle.api.tasks.Nested;
import org.gradle.api.tasks.Optional;
import org.gradle.api.tasks.PathSensitive;
import org.gradle.api.tasks.PathSensitivity;
import org.gradle.api.tasks.TaskAction;
import org.gradle.util.ClosureBackedAction;
import org.gradle.workers.WorkAction;
import org.gradle.workers.WorkQueue;
import org.gradle.workers.WorkerExecutor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public abstract class SpotBugsTask extends DefaultTask
// TODO consider to implements VerificationTask
{
  private static final String FEATURE_FLAG_WORKER_API = "com.github.spotbugs.snom.worker";
  private final Logger log = LoggerFactory.getLogger(SpotBugsTask.class);

  private final WorkerExecutor workerExecutor;

  @NonNull final Property<Boolean> ignoreFailures;
  @NonNull final Property<Boolean> showProgress;
  @NonNull final Property<Confidence> reportLevel;
  @NonNull final Property<Effort> effort;
  @NonNull final ListProperty<String> visitors;
  @NonNull final ListProperty<String> omitVisitors;
  @NonNull final Property<File> reportsDir;
  @NonNull final NamedDomainObjectContainer<SpotBugsReport> reports;

  @NonNull final Property<File> includeFilter;
  @NonNull final Property<File> excludeFilter;
  @NonNull final ListProperty<String> onlyAnalyze;
  @NonNull final Property<String> projectName;
  @NonNull final Property<String> release;

  @InputFiles
  @PathSensitive(PathSensitivity.RELATIVE)
  @NonNull
  public abstract FileCollection getSourceDirs();

  @InputFiles
  @PathSensitive(PathSensitivity.RELATIVE)
  @NonNull
  public abstract FileCollection getClassDirs();

  @InputFiles
  @PathSensitive(PathSensitivity.RELATIVE)
  @NonNull
  public abstract FileCollection getAuxClassPaths();

  @Input
  @Optional
  @NonNull
  public Property<Boolean> getIgnoreFailures() {
    return ignoreFailures;
  }

  @Input
  @Optional
  @NonNull
  public Property<Boolean> getShowProgress() {
    return showProgress;
  }

  @Input
  @Optional
  @NonNull
  public Property<Confidence> getReportLevel() {
    return reportLevel;
  }

  @Input
  @Optional
  @NonNull
  public Property<Effort> getEffort() {
    return effort;
  }

  @Input
  @NonNull
  public ListProperty<String> getVisitors() {
    return visitors;
  }

  @Input
  @NonNull
  public ListProperty<String> getOmitVisitors() {
    return omitVisitors;
  }

  @NonNull
  @Internal("Refer the destination of each report instead.")
  public Property<File> getReportsDir() {
    return reportsDir;
  }

  public void setReportsDir(Provider<File> provider) {
    reportsDir.set(provider);
  }

  @NonNull
  @Optional
  @InputFile
  @PathSensitive(PathSensitivity.RELATIVE)
  public Property<File> getIncludeFilter() {
    return includeFilter;
  }

  @NonNull
  @Optional
  @InputFile
  @PathSensitive(PathSensitivity.RELATIVE)
  public Property<File> getExcludeFilter() {
    return excludeFilter;
  }

  @NonNull
  @Input
  public ListProperty<String> getOnlyAnalyze() {
    return onlyAnalyze;
  }

  public SpotBugsTask(ObjectFactory objects, WorkerExecutor workerExecutor) {
    this.workerExecutor = workerExecutor;

    ignoreFailures = objects.property(Boolean.class);
    showProgress = objects.property(Boolean.class);
    reportLevel = objects.property(Confidence.class);
    effort = objects.property(Effort.class);
    visitors = objects.listProperty(String.class);
    omitVisitors = objects.listProperty(String.class);
    reportsDir = objects.property(File.class);
    reports =
        objects.domainObjectContainer(
            SpotBugsReport.class,
            (name) -> {
              switch (name) {
                case "html":
                  return new SpotBugsHtmlReport(objects, this);
                case "xml":
                  return new SpotBugsXmlReport(objects, this);
                case "text":
                  return new SpotBugsTextReport(objects, this);
                default:
                  throw new InvalidUserDataException(name + " is invalid as the report name");
              }
            });
    includeFilter = objects.property(File.class);
    excludeFilter = objects.property(File.class);
    onlyAnalyze = objects.listProperty(String.class);
    projectName = objects.property(String.class);
    release = objects.property(String.class);
  }

  /**
   * Set properties from extension right after the task creation. User may overwrite these
   * properties by build script.
   *
   * @param extension the source extension to copy the properties.
   */
  @OverrideMustInvoke
  protected void init(SpotBugsExtension extension) {
    ignoreFailures.set(extension.ignoreFailures);
    showProgress.set(extension.showProgress);
    reportLevel.set(extension.reportLevel);
    effort.set(extension.effort);
    visitors.set(extension.visitors);
    omitVisitors.set(extension.omitVisitors);
    // the default reportsDir is "$buildDir/reports/spotbugs/${taskName}"
    reportsDir.set(extension.reportsDir.map(dir -> new File(dir, getName())));
    includeFilter.set(extension.includeFilter);
    excludeFilter.set(extension.excludeFilter);
    onlyAnalyze.set(extension.onlyAnalyze);
    projectName.set(extension.projectName.map(p -> String.format("%s (%s)", p, getName())));
    release.set(extension.release);
  }

  final void applyTo(ImmutableSpotBugsSpec.Builder builder) {
    builder.isIgnoreFailures(ignoreFailures.getOrElse(false));
    builder.isShowProgress(showProgress.getOrElse(false));
    getFirstEnabledReport()
        .ifPresent(
            report -> {
              File dir = report.getDestination().getParentFile();
              dir.mkdirs();
              report.toCommandLineOption().ifPresent(builder::addExtraArguments);
              builder.addExtraArguments("-outputFile", report.getDestination().getAbsolutePath());
            });
    if (effort.isPresent()) {
      builder.addExtraArguments("-effort:" + effort.get().toString().toLowerCase());
    }
    if (reportLevel.isPresent()) {
      builder.addExtraArguments(reportLevel.get().toCommandLineOption());
    }
    if (visitors.isPresent() && !visitors.get().isEmpty()) {
      builder.addExtraArguments("-visitors");
      builder.addExtraArguments(visitors.get().stream().collect(Collectors.joining(",")));
    }
    if (omitVisitors.isPresent() && !omitVisitors.get().isEmpty()) {
      builder.addExtraArguments("-omitVisitors");
      builder.addExtraArguments(omitVisitors.get().stream().collect(Collectors.joining(",")));
    }
    if (includeFilter.isPresent() && includeFilter.get() != null) {
      builder.addExtraArguments("-include", includeFilter.get().getAbsolutePath());
    }
    if (excludeFilter.isPresent() && excludeFilter.get() != null) {
      builder.addExtraArguments("-exclude", excludeFilter.get().getAbsolutePath());
    }
    if (onlyAnalyze.isPresent() && !onlyAnalyze.get().isEmpty()) {
      builder.addExtraArguments(
          "-onlyAnalyze", onlyAnalyze.get().stream().collect(Collectors.joining(",")));
    }

    builder.addExtraArguments("-projectName", projectName.get());
    builder.addExtraArguments("-release", release.getOrElse(getProject().getVersion().toString()));
    builder
        .sourceDirs(getSourceDirs())
        .addAllClassDirs(getClassDirs())
        .addAllAuxClassPaths(getAuxClassPaths());
  }

  @TaskAction
  public void run() {
    ImmutableSpotBugsSpec.Builder builder = ImmutableSpotBugsSpec.builder();
    applyTo(builder);
    Configuration pluginConfig = getProject().getConfigurations().getByName("spotbugsPlugin");
    SpotBugsSpec spotbugsSpec =
        builder
            .addAllSpotbugsJar(createJarOnClasspath())
            .addAllPlugins(pluginConfig.getFiles())
            .build();

    if (getProject().hasProperty(FEATURE_FLAG_WORKER_API)
        && getProject()
            .property(FEATURE_FLAG_WORKER_API)
            .toString()
            .equals(Boolean.TRUE.toString())) {
      log.info("Experimental: Try to run SpotBugs in the worker process.");
      runByWorkerAPI(spotbugsSpec);
    } else {
      runByJavaExec(spotbugsSpec);
    }
  }

  @Nested
  public final NamedDomainObjectContainer<? extends SingleFileReport> getReports() {
    return reports;
  }

  public final NamedDomainObjectContainer<? extends SpotBugsReport> reports(
      Closure<NamedDomainObjectContainer<? extends SpotBugsReport>> closure) {
    return reports(
        new ClosureBackedAction<NamedDomainObjectContainer<? extends SpotBugsReport>>(closure));
  }

  public final NamedDomainObjectContainer<? extends SpotBugsReport> reports(
      Action<NamedDomainObjectContainer<? extends SpotBugsReport>> configureAction) {
    configureAction.execute(reports);
    return reports;
  }

  public void setVisitors(@Nullable Collection<String> collection) {
    visitors.set(collection);
  }

  public void setOmitVisitors(@Nullable Collection<String> collection) {
    omitVisitors.set(collection);
  }

  public void setOnlyAnalyze(Collection<String> collection) {
    onlyAnalyze.set(collection);
  }

  @NonNull
  private Set<File> createJarOnClasspath() {
    Configuration config = getProject().getConfigurations().getByName(SpotBugsPlugin.CONFIG_NAME);
    Configuration spotbugsSlf4j = getProject().getConfigurations().getByName("spotbugsSlf4j");

    Set<File> spotbugsJar = config.getFiles();
    log.info("SpotBugs jar file: {}", spotbugsJar);
    Set<File> slf4jJar = spotbugsSlf4j.getFiles();
    log.info("SLF4J provider jar file: {}", slf4jJar);

    Set<File> jarOnClasspath = new HashSet<>();
    jarOnClasspath.addAll(spotbugsJar);
    jarOnClasspath.addAll(slf4jJar);
    return jarOnClasspath;
  }

  @NonNull
  @Nested
  public java.util.Optional<SpotBugsReport> getFirstEnabledReport() {
    return reports.stream().filter(SingleFileReport::isEnabled).findFirst();
  }

  /**
   * Run SpotBugs in the worker process.
   *
   * @param spotbugsSpec Spec to run SpotBugs
   */
  private void runByWorkerAPI(SpotBugsSpec spotbugsSpec) {
    WorkQueue workerQueue = workerExecutor.processIsolation(spotbugsSpec::applyTo);
    workerQueue.submit(SpotBugsExecutor.class, spotbugsSpec::applyTo);
  }

  private void runByJavaExec(SpotBugsSpec spotbugsSpec) {
    // TODO print version of SpotBugs and Plugins
    getProject().javaexec(spotbugsSpec::applyTo);
    // TODO handle isIgnoreFailures
  }

  public abstract static class SpotBugsExecutor implements WorkAction<SpotBugsWorkParameters> {
    @Override
    public void execute() {
      SpotBugsWorkParameters params = getParameters();
      String[] args = params.getArguments().get().toArray(new String[0]);

      // TODO handle isIgnoreFailures
      try {
        edu.umd.cs.findbugs.Version.printVersion(false);
        edu.umd.cs.findbugs.FindBugs2.main(args);
      } catch (Exception e) {
        throw new GradleException("SpotBugs execution thrown exception", e);
      }
    }
  }
}

File: src/main/kotlin/com/github/spotbugs/snom/SpotBugsTask.kt
/*
 * Copyright 2021 SpotBugs team
 *
 * <p>Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
 * except in compliance with the License. You may obtain a copy of the License at
 *
 * <p>http://www.apache.org/licenses/LICENSE-2.0
 *
 * <p>Unless required by applicable law or agreed to in writing, software distributed under the
 * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.spotbugs.snom

import com.github.spotbugs.snom.internal.SpotBugsHtmlReport
import com.github.spotbugs.snom.internal.SpotBugsRunnerForHybrid
import com.github.spotbugs.snom.internal.SpotBugsRunnerForJavaExec
import com.github.spotbugs.snom.internal.SpotBugsSarifReport
import com.github.spotbugs.snom.internal.SpotBugsTextReport
import com.github.spotbugs.snom.internal.SpotBugsXmlReport
import javax.inject.Inject
import org.gradle.api.Action
import org.gradle.api.DefaultTask
import org.gradle.api.InvalidUserDataException
import org.gradle.api.NamedDomainObjectContainer
import org.gradle.api.file.ConfigurableFileCollection
import org.gradle.api.file.DirectoryProperty
import org.gradle.api.file.FileCollection
import org.gradle.api.file.RegularFileProperty
import org.gradle.api.plugins.JavaBasePlugin
import org.gradle.api.plugins.JavaPluginExtension
import org.gradle.api.provider.ListProperty
import org.gradle.api.provider.Property
import org.gradle.api.tasks.CacheableTask
import org.gradle.api.tasks.Classpath
import org.gradle.api.tasks.Input
import org.gradle.api.tasks.InputFile
import org.gradle.api.tasks.InputFiles
import org.gradle.api.tasks.Internal
import org.gradle.api.tasks.Nested
import org.gradle.api.tasks.Optional
import org.gradle.api.tasks.OutputFile
import org.gradle.api.tasks.PathSensitive
import org.gradle.api.tasks.PathSensitivity
import org.gradle.api.tasks.SkipWhenEmpty
import org.gradle.api.tasks.TaskAction
import org.gradle.api.tasks.VerificationTask
import org.gradle.jvm.toolchain.JavaLauncher
import org.gradle.jvm.toolchain.JavaToolchainService
import org.gradle.process.ExecOperations
import org.gradle.workers.WorkerExecutor
import org.slf4j.LoggerFactory

/**
 * This Gradle task runs the SpotBugs analysis.
 * All properties are optional.
 *
 * **Usage for Java projects:**
 * After applying the SpotBugs Gradle plugin to your project, a [SpotBugsTask] is automatically
 * generated for each sourceSet.
 * If you need to configure the generated tasks, use the following build script as a guide:
 *
 * ```groovy
 * spotbugsMain {
 *     sourceSets = sourceSets.main.allSource.srcDirs
 *     classDirs = sourceSets.main.output
 *     auxClassPaths = sourceSets.main.compileClasspath
 *     ignoreFailures = false
 *     showStackTraces = true
 *     showProgress = false
 *     reportLevel = 'default'
 *     effort = 'default'
 *     visitors = [ 'FindSqlInjection', 'SwitchFallthrough' ]
 *     omitVisitors = [ 'FindNonShortCircuit' ]
 *     reportsDir = file("$buildDir/reports/spotbugs")
 *     includeFilter = file('spotbugs-include.xml')
 *     excludeFilter = file('spotbugs-exclude.xml')
 *     baselineFile = file('spotbugs-baseline.xml')
 *     onlyAnalyze = ['com.foobar.MyClass', 'com.foobar.mypkg.*']
 *     projectName = name
 *     release = version
 *     extraArgs = [ '-nested:false' ]
 *     jvmArgs = [ '-Duser.language=ja' ]
 *     maxHeapSize = '512m'
 * ```
 *
 * See also [SpotBugs Manual about configuration](https://spotbugs.readthedocs.io/en/stable/running.html).
 */
@CacheableTask
abstract class SpotBugsTask : DefaultTask(), VerificationTask {
    private val log = LoggerFactory.getLogger(SpotBugsTask::class.java)

    @get:Inject
    abstract val workerExecutor: WorkerExecutor

    @get:Inject
    abstract val execOps: ExecOperations

    @Input
    override fun getIgnoreFailures(): Boolean = ignoreFailures.get()

    override fun setIgnoreFailures(ignoreFailures: Boolean) {
        this.ignoreFailures.set(ignoreFailures)
    }

    private val ignoreFailures = project.objects.property(Boolean::class.java)

    @get:Input
    abstract val showStackTraces: Property<Boolean>

    /**
     * Property to enable progress reporting during the analysis. Default value is `false`.
     */
    @get:Optional
    @get:Input
    abstract val showProgress: Property<Boolean>

    /**
     * Property to specify the level to report bugs. Default value is [Confidence.DEFAULT].
     */
    @get:Input
    @get:Optional
    abstract val reportLevel: Property<Confidence>

    /**
     * Property to adjust SpotBugs detectors. Default value is [Effort.DEFAULT].
     */
    @get:Input
    @get:Optional
    abstract val effort: Property<Effort>

    /**
     * Property to enable visitors (detectors) for analysis. Default is empty that means all visitors run analysis.
     */
    @get:Input
    abstract val visitors: ListProperty<String>

    /**
     * Property to disable visitors (detectors) for analysis. Default is empty that means SpotBugs omits no visitor.
     */
    @get:Input
    abstract val omitVisitors: ListProperty<String>

    /**
     * Property to set the directory to generate report files. Default is `"$buildDir/reports/spotbugs/$taskName"}`.
     */
    @get:Internal("Refer the destination of each report instead.")
    abstract val reportsDir: DirectoryProperty

    /**
     * Property defined to keep the backward compatibility with [org.gradle.api.reporting.Reporting] interface.
     *
     * See [SpotBugsReport]
     */
    @get:Internal
    val reports: NamedDomainObjectContainer<SpotBugsReport>

    /**
     * Property to set the filter file to limit which bug should be reported.
     *
     * Note that this property will NOT limit which bug should be detected. To limit the target classes to analyze,
     * use [onlyAnalyze] instead.
     * To limit the visitors (detectors) to run, use [visitors] and [omitVisitors] instead.
     *
     * See also [SpotBugs Manual about Filter file](https://spotbugs.readthedocs.io/en/stable/filter.html).
     */
    @get:Optional
    @get:InputFile
    @get:PathSensitive(PathSensitivity.RELATIVE)
    abstract val includeFilter: RegularFileProperty

    /**
     * Property to set the filter file to limit which bug should be reported.
     *
     * Note that this property will NOT limit which bug should be detected. To limit the target classes to analyze,
     * use [onlyAnalyze] instead.
     * To limit the visitors (detectors) to run, use [visitors] and [omitVisitors] instead.
     *
     * See also [SpotBugs Manual about Filter file](https://spotbugs.readthedocs.io/en/stable/filter.html).
     */
    @get:Optional
    @get:InputFile
    @get:PathSensitive(PathSensitivity.RELATIVE)
    abstract val excludeFilter: RegularFileProperty

    /**
     * Property to set the baseline file. This file is a Spotbugs result file, and all bugs reported in this file
     * will not be reported in the final output.
     */
    @get:Optional
    @get:InputFile
    @get:PathSensitive(PathSensitivity.RELATIVE)
    abstract val baselineFile: RegularFileProperty

    /**
     * Property to specify the target classes for analysis. Default value is empty that means all classes are analyzed.
     */
    @get:Input
    abstract val onlyAnalyze: ListProperty<String>

    /**
     * Property to specify the name of project. Some reporting formats use this property.
     * Default value is `"${project.name} (${task.name})"`.
     *
     * Note that this property, if treated as a task input, can break cacheability.
     * As such, it has been marked [Internal] to exclude it from task up-to-date and
     * cacheability checks.
     */
    @get:Internal
    abstract val projectName: Property<String>

    /**
     * Property to specify the release identifier of project. Some reporting formats use this property.
     * Default value is the version of your Gradle project.
     */
    @get:Input
    abstract val release: Property<String>

    /**
     * Property to specify the extra arguments for SpotBugs. Default value is empty so SpotBugs will get no
     * extra argument.
     */
    @get:Optional
    @get:Input
    abstract val extraArgs: ListProperty<String>

    /**
     * Property to specify the extra arguments for JVM process. Default value is empty so JVM process will get no
     * extra argument.
     */
    @get:Optional
    @get:Input
    abstract val jvmArgs: ListProperty<String>

    /**
     * Property to specify the max heap size (`-Xmx` option) of JVM process.
     * Default value is empty so the default configuration made by Gradle will be used.
     */
    @get:Optional
    @get:Input
    abstract val maxHeapSize: Property<String>

    /**
     * Property to specify the directories that contain the source of target classes to analyze.
     * Default value is the source directory of the target sourceSet.
     */
    @get:InputFiles
    @get:PathSensitive(PathSensitivity.RELATIVE)
    abstract val sourceDirs: ConfigurableFileCollection

    /**
     * Property to specify the directories that contains the target classes to analyze.
     * Default value is the output directory of the target sourceSet.
     */
    @get:Internal
    abstract val classDirs: ConfigurableFileCollection

    /**
     * Property to specify the aux class paths that contains the libraries to refer during analysis.
     * Default value is the compile-scope dependencies of the target sourceSet.
     */
    @get:Classpath
    abstract val auxClassPaths: ConfigurableFileCollection

    /**
     * Property to enable auxclasspathFromFile and prevent Argument List Too Long issues in java processes.
     * Default value is `false`.
     */
    @get:Input
    @get:Optional
    abstract val useAuxclasspathFile: Property<Boolean>

    @get:Internal
    abstract val auxclasspathFile: RegularFileProperty

    /**
     * Property to specify the target classes to analyze by SpotBugs.
     * Default value is the all existing `.class` files in `getClassDirs`.
     */
    @get:InputFiles
    @get:PathSensitive(PathSensitivity.RELATIVE)
    @get:SkipWhenEmpty
    var classes: FileCollection? = null
        get() {
            return field ?: classDirs.asFileTree.filter {
                it.name.endsWith(".class")
            }
        }

    private var enableWorkerApi: Boolean = true

    @get:Internal
    abstract val pluginJarFiles: ConfigurableFileCollection

    @get:Internal
    abstract val spotbugsClasspath: ConfigurableFileCollection

    @get:Nested
    @get:Optional
    abstract val launcher: Property<JavaLauncher>

    /**
     * A file that lists class files and jar files to analyze.
     */
    @get:OutputFile
    abstract val analyseClassFile: RegularFileProperty

    init {
        val objects = project.objects
        reports = objects.domainObjectContainer(SpotBugsReport::class.java) { name: String ->
            when (name) {
                "html" -> objects.newInstance(SpotBugsHtmlReport::class.java, name, objects, this)
                "xml" -> objects.newInstance(SpotBugsXmlReport::class.java, name, objects, this)
                "text" -> objects.newInstance(SpotBugsTextReport::class.java, name, objects, this)
                "sarif" -> objects.newInstance(SpotBugsSarifReport::class.java, name, objects, this)
                else -> throw InvalidUserDataException("$name is invalid as the report name")
            }.also {
                (outputs as org.gradle.api.tasks.TaskOutputs).file(it.outputLocation)
            }
        }
        description = "Run SpotBugs analysis."
        group = JavaBasePlugin.VERIFICATION_GROUP
    }

    /**
     * Set properties from extension right after the task creation. User may overwrite these
     * properties by build script.
     *
     * @param extension the source extension to copy the properties.
     */
    fun init(
        extension: SpotBugsExtension,
        enableWorkerApi: Boolean,
    ) {
        auxclasspathFile.convention(project.layout.buildDirectory.file("spotbugs/auxclasspath/$name"))
        ignoreFailures.convention(extension.ignoreFailures)
        showStackTraces.convention(extension.showStackTraces)
        showProgress.convention(extension.showProgress)
        reportLevel.convention(extension.reportLevel)
        effort.convention(extension.effort)
        visitors.convention(extension.visitors)
        omitVisitors.convention(extension.omitVisitors)
        // the default reportsDir is "$buildDir/reports/spotbugs/"
        reportsDir.convention(extension.reportsDir)
        includeFilter.convention(extension.includeFilter)
        excludeFilter.convention(extension.excludeFilter)
        baselineFile.convention(extension.baselineFile)
        onlyAnalyze.convention(extension.onlyAnalyze)
        projectName.convention(extension.projectName.map { p -> "$p ($name)" })
        release.convention(extension.release)
        jvmArgs.convention(extension.jvmArgs)
        extraArgs.convention(extension.extraArgs)
        maxHeapSize.convention(extension.maxHeapSize)
        useAuxclasspathFile.convention(extension.useAuxclasspathFile)

        if (extension.useJavaToolchains.isPresent && extension.useJavaToolchains.get()) {
            configureJavaLauncher()
        }

        this.enableWorkerApi = enableWorkerApi

        analyseClassFile.set(project.layout.buildDirectory.file("$name-analyse-class-file.txt"))

        val pluginConfiguration = project.configurations.getByName(SpotBugsPlugin.PLUGINS_CONFIG_NAME)
        pluginJarFiles.from(
            project.provider { pluginConfiguration.files },
        )
        val configuration = project.configurations.getByName(SpotBugsPlugin.CONFIG_NAME)
        val spotbugsSlf4j = project.configurations.getByName(SpotBugsPlugin.SLF4J_CONFIG_NAME)
        spotbugsClasspath.from(
            project.layout.files(
                project.provider { spotbugsSlf4j.files },
                project.provider { configuration.files },
            ),
        )
    }

    /**
     * Set convention for default java launcher based on Toolchain configuration
     */
    private fun configureJavaLauncher() {
        val toolchain = project.extensions.getByType(JavaPluginExtension::class.java).toolchain
        val service = project.extensions.getByType(JavaToolchainService::class.java)
        val defaultLauncher = service.launcherFor(toolchain)
        launcher.convention(defaultLauncher)
    }

    @TaskAction
    fun run() {
        if (enableWorkerApi) {
            log.info("Running SpotBugs by Gradle no-isolated Worker...")
            SpotBugsRunnerForHybrid(workerExecutor, launcher).run(this)
        } else {
            log.info("Running SpotBugs by JavaExec...")
            SpotBugsRunnerForJavaExec(execOps, launcher).run(this)
        }
    }

    /**
     * Function defined to keep the backward compatibility with [org.gradle.api.reporting.Reporting] interface.
     */
    @Suppress("MaxLineLength")
    fun reports(configureAction: Action<NamedDomainObjectContainer<SpotBugsReport>>): NamedDomainObjectContainer<SpotBugsReport> {
        configureAction.execute(reports)
        return reports
    }

    @Internal
    fun getBaseName(): String {
        var prunedName = name.replaceFirst("spotbugs", "")
        if (prunedName.isEmpty()) {
            prunedName = name
        }

        return buildString {
            append(Character.toLowerCase(prunedName[0]))
            append(prunedName.substring(1))
        }
    }

    @Internal
    internal fun getRequiredReports(): Sequence<SpotBugsReport> {
        return reports.matching { it.required.get() }.asMap.values.asSequence()
    }
}
